File: src/main/storage/sandbox.ts
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  realpathSync,
  renameSync,
  rmSync,
  statSync,
  writeFileSync
} from 'fs';
import { readdir, stat } from 'fs/promises';
import { basename, dirname, isAbsolute, join, relative, resolve } from 'path';

interface IncomingResumeMeta {
  fileId: string;
  deviceId: string;
  deviceName: string;
  trustFingerprint: string;
  trustPublicKey: string;
  fileName: string;
  fileSize: number;
  sha256: string;
  finalPath: string;
  partialPath: string;
}

export interface ResumeCacheSummary {
  count: number;
  bytes: number;
}

export interface ResumeCacheEntry {
  fileId: string;
  deviceId: string;
  deviceName: string;
  trustFingerprint: string;
  trustPublicKey: string;
  fileName: string;
  fileSize: number;
  sha256: string;
  partialPath: string;
  finalPath: string;
  bytesReceived: number;
}

export class Sandbox {
  private usageBytes: number | null = null;
  private usageRefreshPromise: Promise<number> | null = null;
  private usageDirty = true;
  private usageDirtyWhileRefreshing = false;

  constructor(private root: string) {}

  setRoot(root: string): void {
    this.root = root;
    this.markUsageDirty();
  }

  rootPath(): string {
    mkdirSync(this.root, { recursive: true });
    return this.root;
  }

  containsPath(targetPath: string): boolean {
    return isPathWithinRoot(this.rootPath(), targetPath);
  }

  assertContainsPath(targetPath: string): string {
    if (!this.containsPath(targetPath)) {
      throw new Error('path is outside sandbox');
    }
    return targetPath;
  }

  async currentUsageBytes(): Promise<number> {
    if (!this.usageDirty && this.usageBytes !== null) {
      return this.usageBytes;
    }

    return this.refreshUsageBytes();
  }

  markUsageDirty(): void {
    this.usageDirty = true;
    if (this.usageRefreshPromise) {
      this.usageDirtyWhileRefreshing = true;
    }
  }

  directoryForIncoming(deviceId: string): string {
    const safeDeviceId = sanitizeSegment(deviceId);
    const deviceDir = join(this.root, safeDeviceId);
    mkdirSync(deviceDir, { recursive: true });
    return deviceDir;
  }

  pathForIncoming(deviceId: string, originalFileName: string): string {
    const deviceDir = this.directoryForIncoming(deviceId);
    const safeName = sanitizeSegment(basename(originalFileName));
    const stamp = formatTimestamp(new Date());
    return join(deviceDir, `${stamp}_${safeName}`);
  }

  prepareIncomingResume(
    fileId: string,
    deviceId: string,
    deviceName: string,
    trustFingerprint: string,
    trustPublicKey: string,
    fileName: string,
    fileSize: number,
    sha256: string
  ): {
    finalPath: string;
    partialPath: string;
    bytesReceived: number;
  } {
    const existing = this.readIncomingResumeMeta(fileId);
    if (
      existing &&
      isMatchingIncomingResume(existing, {
        deviceId,
        deviceName,
        trustFingerprint,
        trustPublicKey,
        fileName,
        fileSize,
        sha256
      }) &&
      existsSync(existing.partialPath)
    ) {
      return {
        finalPath: existing.finalPath,
        partialPath: existing.partialPath,
        bytesReceived: statSync(existing.partialPath).size
      };
    }

    const finalPath = this.pathForIncoming(deviceId, fileName);
    const partialPath = `${finalPath}.part`;
    const meta: IncomingResumeMeta = {
      fileId,
      deviceId,
      deviceName,
      trustFingerprint,
      trustPublicKey,
      fileName,
      fileSize,
      sha256,
      finalPath,
      partialPath
    };
    this.writeIncomingResumeMeta(meta);

    return {
      finalPath,
      partialPath,
      bytesReceived: 0
    };
  }

  completeIncomingResume(fileId: string): string {
    const meta = this.readIncomingResumeMeta(fileId);
    if (!meta) {
      throw new Error(`resume state ${fileId} not found`);
    }
    renameSync(meta.partialPath, meta.finalPath);
    rmSync(this.resumeMetaPath(fileId), { force: true });
    this.markUsageDirty();
    return meta.finalPath;
  }

  discardIncomingResume(fileId: string, removePartial: boolean): void {
    const meta = this.readIncomingResumeMeta(fileId);
    if (!meta) {
      return;
    }
    if (removePartial) {
      rmSync(meta.partialPath, { force: true });
    }
    rmSync(this.resumeMetaPath(fileId), { force: true });
    this.markUsageDirty();
  }

  incomingResumeOffset(fileId: string): number {
    const meta = this.readIncomingResumeMeta(fileId);
    if (!meta || !existsSync(meta.partialPath)) {
      return 0;
    }
    return statSync(meta.partialPath).size;
  }

  matchingResumeBytes(
    fileId: string,
    deviceId: string,
    deviceName: string,
    trustFingerprint: string,
    trustPublicKey: string,
    fileName: string,
    fileSize: number,
    sha256: string
  ): number {
    const meta = this.readIncomingResumeMeta(fileId);
    if (
      !meta ||
      !isMatchingIncomingResume(meta, {
        deviceId,
        deviceName,
        trustFingerprint,
        trustPublicKey,
        fileName,
        fileSize,
        sha256
      }) ||
      !existsSync(meta.partialPath)
    ) {
      return 0;
    }

    return statSync(meta.partialPath).size;
  }

  hasIncomingResume(fileId: string): boolean {
    const meta = this.readIncomingResumeMeta(fileId);
    return Boolean(meta && existsSync(meta.partialPath));
  }

  resumeCacheSummary(): ResumeCacheSummary {
    const entries = this.listResumeEntries();
    return {
      count: entries.length,
      bytes: entries.reduce((total, entry) => total + entry.bytesReceived, 0)
    };
  }

  clearResumeCache(excludedFileIds: Set<string> = new Set()): string[] {
    const cleared: string[] = [];
    for (const entry of this.listResumeEntries()) {
      if (excludedFileIds.has(entry.fileId)) {
        continue;
      }
      this.discardIncomingResume(entry.fileId, true);
      cleared.push(entry.fileId);
    }
    return cleared;
  }

  listResumeEntries(): ResumeCacheEntry[] {
    const resumeDir = this.resumeDirectoryPath();
    if (!existsSync(resumeDir)) {
      return [];
    }

    const entries: ResumeCacheEntry[] = [];
    for (const entry of readdirSync(resumeDir, { withFileTypes: true })) {
      if (!entry.isFile() || !entry.name.endsWith('.json')) {
        continue;
      }
      const fileId = entry.name.slice(0, -5);
      const meta = this.readIncomingResumeMeta(fileId);
      if (!meta || !existsSync(meta.partialPath)) {
        continue;
      }
      entries.push({
        fileId: meta.fileId,
        deviceId: meta.deviceId,
        deviceName: meta.deviceName,
        trustFingerprint: meta.trustFingerprint,
        trustPublicKey: meta.trustPublicKey,
        fileName: meta.fileName,
        fileSize: meta.fileSize,
        sha256: meta.sha256,
        partialPath: meta.partialPath,
        finalPath: meta.finalPath,
        bytesReceived: statSync(meta.partialPath).size
      });
    }

    return entries;
  }

  private resumeMetaPath(fileId: string): string {
    const resumeDir = this.resumeDirectoryPath();
    return join(resumeDir, `${sanitizeSegment(fileId)}.json`);
  }

  private resumeDirectoryPath(): string {
    const resumeDir = join(this.rootPath(), '.resume');
    mkdirSync(resumeDir, { recursive: true });
    return resumeDir;
  }

  private readIncomingResumeMeta(fileId: string): IncomingResumeMeta | null {
    const metaPath = this.resumeMetaPath(fileId);
    if (!existsSync(metaPath)) {
      return null;
    }

    try {
      const raw = readFileSync(metaPath, 'utf8');
      const parsed = JSON.parse(raw) as Partial<IncomingResumeMeta>;
      if (
        typeof parsed.fileId !== 'string' ||
        typeof parsed.deviceId !== 'string' ||
        typeof parsed.fileName !== 'string' ||
        typeof parsed.fileSize !== 'number' ||
        typeof parsed.sha256 !== 'string' ||
        typeof parsed.finalPath !== 'string' ||
        typeof parsed.partialPath !== 'string'
      ) {
        return null;
      }
      const normalized: IncomingResumeMeta = {
        fileId: parsed.fileId,
        deviceId: parsed.deviceId,
        deviceName: typeof parsed.deviceName === 'string' && parsed.deviceName.length > 0 ? parsed.deviceName : parsed.deviceId,
        trustFingerprint:
          typeof parsed.trustFingerprint === 'string' && parsed.trustFingerprint.length > 0
            ? parsed.trustFingerprint
            : '',
        trustPublicKey:
          typeof parsed.trustPublicKey === 'string' && parsed.trustPublicKey.length > 0
            ? parsed.trustPublicKey
            : '',
        fileName: parsed.fileName,
        fileSize: parsed.fileSize,
        sha256: parsed.sha256,
        finalPath: parsed.finalPath,
        partialPath: parsed.partialPath
      };
      return normalized;
    } catch {
      return null;
    }
  }

  private writeIncomingResumeMeta(meta: IncomingResumeMeta): void {
    writeFileSync(this.resumeMetaPath(meta.fileId), JSON.stringify(meta, null, 2), 'utf8');
    this.markUsageDirty();
  }

  private refreshUsageBytes(): Promise<number> {
    if (this.usageRefreshPromise) {
      return this.usageRefreshPromise;
    }

    this.usageDirtyWhileRefreshing = false;
    this.usageRefreshPromise = directorySize(this.rootPath())
      .then((totalBytes) => {
        this.usageBytes = totalBytes;
        this.usageDirty = this.usageDirtyWhileRefreshing;
        return totalBytes;
      })
      .finally(() => {
        this.usageRefreshPromise = null;
      });

    return this.usageRefreshPromise;
  }
}

function sanitizeSegment(input: string): string {
  return input.replace(/[\\/\0]/g, '_').replace(/^\.+/, '') || 'unnamed';
}

function isMatchingIncomingResume(
  existing: IncomingResumeMeta,
  candidate: {
    deviceId: string;
    deviceName: string;
    trustFingerprint: string;
    trustPublicKey: string;
    fileName: string;
    fileSize: number;
    sha256: string;
  }
): boolean {
  return (
    existing.deviceId === candidate.deviceId &&
    existing.deviceName === candidate.deviceName &&
    existing.trustFingerprint === candidate.trustFingerprint &&
    existing.trustPublicKey === candidate.trustPublicKey &&
    existing.fileName === candidate.fileName &&
    existing.fileSize === candidate.fileSize &&
    existing.sha256 === candidate.sha256
  );
}

function formatTimestamp(d: Date): string {
  const yyyy = d.getFullYear().toString();
  const mm = String(d.getMonth() + 1).padStart(2, '0');
  const dd = String(d.getDate()).padStart(2, '0');
  const hh = String(d.getHours()).padStart(2, '0');
  const mi = String(d.getMinutes()).padStart(2, '0');
  const ss = String(d.getSeconds()).padStart(2, '0');
  return `${yyyy}${mm}${dd}_${hh}${mi}${ss}`;
}

async function directorySize(path: string): Promise<number> {
  let total = 0;

  for (const entry of await readdir(path, { withFileTypes: true })) {
    const fullPath = join(path, entry.name);
    if (entry.isDirectory()) {
      total += await directorySize(fullPath);
      continue;
    }
    if (entry.isFile()) {
      total += (await stat(fullPath)).size;
    }
  }

  return total;
}

function isPathWithinRoot(rootPath: string, targetPath: string): boolean {
  const normalizedRoot = normalizePathForComparison(rootPath);
  const normalizedTarget = normalizePathForComparison(targetPath);
  const relativePath = relative(normalizedRoot, normalizedTarget);
  return relativePath === '' || (!relativePath.startsWith('..') && !isAbsolute(relativePath));
}

function normalizePathForComparison(path: string): string {
  const missingSegments: string[] = [];
  let resolvedPath = resolve(path);

  while (!existsSync(resolvedPath)) {
    const parent = dirname(resolvedPath);
    if (parent === resolvedPath) {
      return resolvedPath;
    }
    missingSegments.unshift(basename(resolvedPath));
    resolvedPath = parent;
  }

  try {
    const normalizedPath = realpathSync.native(resolvedPath);
    return missingSegments.length > 0 ? join(normalizedPath, ...missingSegments) : normalizedPath;
  } catch {
    const normalizedPath = realpathSync(resolvedPath);
    return missingSegments.length > 0 ? join(normalizedPath, ...missingSegments) : normalizedPath;
  }
}

File: src-tauri/src/storage/sandbox.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, RwLock};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Clone)]
pub struct Sandbox {
    root: Arc<RwLock<PathBuf>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IncomingResumeMeta {
    pub file_id: String,
    pub device_id: String,
    pub device_name: String,
    pub trust_fingerprint: String,
    pub trust_public_key: String,
    pub file_name: String,
    pub file_size: u64,
    pub sha256: String,
    pub final_path: String,
    pub partial_path: String,
}

#[derive(Debug, Clone)]
pub struct IncomingResumeInfo {
    pub partial_path: PathBuf,
    pub final_path: PathBuf,
    pub bytes_received: u64,
}

#[derive(Debug, Clone, Copy)]
pub struct IncomingResumeRequest<'a> {
    pub file_id: &'a str,
    pub device_id: &'a str,
    pub device_name: &'a str,
    pub trust_fingerprint: &'a str,
    pub trust_public_key: &'a str,
    pub file_name: &'a str,
    pub file_size: u64,
    pub sha256: &'a str,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResumeCacheEntry {
    pub file_id: String,
    pub device_id: String,
    pub device_name: String,
    pub trust_fingerprint: String,
    pub trust_public_key: String,
    pub file_name: String,
    pub file_size: u64,
    pub sha256: String,
    pub partial_path: String,
    pub final_path: String,
    pub bytes_received: u64,
}

impl Sandbox {
    pub fn new(root: PathBuf) -> Self {
        let _ = std::fs::create_dir_all(&root);
        Self {
            root: Arc::new(RwLock::new(root)),
        }
    }

    pub fn root_path(&self) -> PathBuf {
        self.root.read().unwrap().clone()
    }

    pub fn set_root(&self, root: PathBuf) {
        let _ = std::fs::create_dir_all(&root);
        *self.root.write().unwrap() = root;
    }

    pub fn resolve_path(&self, device_id: &str, original_name: &str) -> PathBuf {
        let device_dir = self.root_path().join(device_id);
        let _ = std::fs::create_dir_all(&device_dir);

        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        let sanitized = original_name.replace(['/', '\\'], "_");

        let filename = format!("{}_{}", timestamp, sanitized);
        device_dir.join(filename)
    }

    pub fn ensure_device_dir(&self, device_id: &str) -> PathBuf {
        let dir = self.root_path().join(device_id);
        let _ = std::fs::create_dir_all(&dir);
        dir
    }

    pub fn directory_for_incoming(&self, device_id: &str) -> PathBuf {
        let safe_id = sanitize_segment(device_id);
        let dir = self.root_path().join(&safe_id);
        let _ = std::fs::create_dir_all(&dir);
        dir
    }

    pub fn path_for_incoming(&self, device_id: &str, original_file_name: &str) -> PathBuf {
        let device_dir = self.directory_for_incoming(device_id);
        let safe_name = sanitize_segment(
            std::path::Path::new(original_file_name)
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unnamed"),
        );
        let stamp = format_timestamp(SystemTime::now());
        device_dir.join(format!("{}_{}", stamp, safe_name))
    }

    pub fn prepare_incoming_resume(
        &self,
        request: IncomingResumeRequest<'_>,
    ) -> IncomingResumeInfo {
        if let Some(existing) = self.read_incoming_resume_meta(request.file_id) {
            if is_matching_incoming_resume(&existing, &request) {
                if let Ok(meta) = std::fs::metadata(&existing.partial_path) {
                    return IncomingResumeInfo {
                        partial_path: PathBuf::from(&existing.partial_path),
                        final_path: PathBuf::from(&existing.final_path),
                        bytes_received: meta.len(),
                    };
                }
            }
        }

        let final_path = self.path_for_incoming(request.device_id, request.file_name);
        let partial_path = final_path.with_extension("part");
        let meta = IncomingResumeMeta {
            file_id: request.file_id.to_string(),
            device_id: request.device_id.to_string(),
            device_name: request.device_name.to_string(),
            trust_fingerprint: request.trust_fingerprint.to_string(),
            trust_public_key: request.trust_public_key.to_string(),
            file_name: request.file_name.to_string(),
            file_size: request.file_size,
            sha256: request.sha256.to_string(),
            final_path: final_path.to_string_lossy().to_string(),
            partial_path: partial_path.to_string_lossy().to_string(),
        };
        self.write_incoming_resume_meta(&meta);

        IncomingResumeInfo {
            partial_path,
            final_path,
            bytes_received: 0,
        }
    }

    pub fn complete_incoming_resume(&self, file_id: &str) -> Result<PathBuf, String> {
        let meta = self
            .read_incoming_resume_meta(file_id)
            .ok_or_else(|| format!("resume state {} not found", file_id))?;
        fs::rename(&meta.partial_path, &meta.final_path)
            .map_err(|e| format!("failed to rename: {}", e))?;
        let _ = fs::remove_file(self.resume_meta_path(file_id));
        Ok(PathBuf::from(&meta.final_path))
    }

    pub fn discard_incoming_resume(&self, file_id: &str, remove_partial: bool) {
        if let Some(meta) = self.read_incoming_resume_meta(file_id) {
            if remove_partial {
                let _ = fs::remove_file(&meta.partial_path);
            }
            let _ = fs::remove_file(self.resume_meta_path(file_id));
        }
    }

    pub fn incoming_resume_offset(&self, file_id: &str) -> u64 {
        let meta = match self.read_incoming_resume_meta(file_id) {
            Some(m) => m,
            None => return 0,
        };
        if let Ok(m) = std::fs::metadata(&meta.partial_path) {
            m.len()
        } else {
            0
        }
    }

    pub fn has_incoming_resume(&self, file_id: &str) -> bool {
        self.read_incoming_resume_meta(file_id).is_some()
    }

    pub fn list_resume_entries(&self) -> Vec<ResumeCacheEntry> {
        let resume_dir = self.resume_directory_path();
        if !resume_dir.exists() {
            return Vec::new();
        }

        let mut entries = Vec::new();
        if let Ok(dir) = std::fs::read_dir(resume_dir) {
            for entry in dir.flatten() {
                let path = entry.path();
                if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
                    continue;
                }
                let file_id = match path.file_stem().and_then(|stem| stem.to_str()) {
                    Some(id) => id,
                    None => continue,
                };
                let Some(meta) = self.read_incoming_resume_meta(file_id) else {
                    continue;
                };
                let Ok(partial_meta) = std::fs::metadata(&meta.partial_path) else {
                    continue;
                };
                entries.push(ResumeCacheEntry {
                    file_id: meta.file_id,
                    device_id: meta.device_id,
                    device_name: meta.device_name,
                    trust_fingerprint: meta.trust_fingerprint,
                    trust_public_key: meta.trust_public_key,
                    file_name: meta.file_name,
                    file_size: meta.file_size,
                    sha256: meta.sha256,
                    partial_path: meta.partial_path,
                    final_path: meta.final_path,
                    bytes_received: partial_meta.len(),
                });
            }
        }

        entries
    }

    pub fn clear_resume_cache(&self, excluded_file_ids: &HashSet<String>) -> Vec<String> {
        let mut cleared = Vec::new();
        for entry in self.list_resume_entries() {
            if excluded_file_ids.contains(&entry.file_id) {
                continue;
            }
            self.discard_incoming_resume(&entry.file_id, true);
            cleared.push(entry.file_id);
        }
        cleared
    }

    pub fn current_usage_bytes(&self) -> u64 {
        dir_size(&self.root_path())
    }

    fn resume_meta_path(&self, file_id: &str) -> PathBuf {
        let resume_dir = self.resume_directory_path();
        let _ = std::fs::create_dir_all(&resume_dir);
        resume_dir.join(format!("{}.json", sanitize_segment(file_id)))
    }

    fn resume_directory_path(&self) -> PathBuf {
        self.root_path().join(".resume")
    }

    fn read_incoming_resume_meta(&self, file_id: &str) -> Option<IncomingResumeMeta> {
        let path = self.resume_meta_path(file_id);
        let content = std::fs::read_to_string(&path).ok()?;
        serde_json::from_str(&content).ok()
    }

    fn write_incoming_resume_meta(&self, meta: &IncomingResumeMeta) {
        let path = self.resume_meta_path(&meta.file_id);
        if let Ok(json) = serde_json::to_string_pretty(meta) {
            let _ = std::fs::write(path, json);
        }
    }
}

fn sanitize_segment(input: &str) -> String {
    let result: String = input
        .chars()
        .map(|c| match c {
            '/' | '\\' | '\0' => '_',
            c => c,
        })
        .collect();
    let trimmed = result.trim_start_matches('.');
    if trimmed.is_empty() {
        "unnamed".to_string()
    } else {
        trimmed.to_string()
    }
}

fn is_matching_incoming_resume(
    existing: &IncomingResumeMeta,
    request: &IncomingResumeRequest<'_>,
) -> bool {
    existing.device_id == request.device_id
        && existing.device_name == request.device_name
        && existing.trust_fingerprint == request.trust_fingerprint
        && existing.trust_public_key == request.trust_public_key
        && existing.file_name == request.file_name
        && existing.file_size == request.file_size
        && existing.sha256 == request.sha256
}

fn format_timestamp(time: SystemTime) -> String {
    let dur = time.duration_since(UNIX_EPOCH).unwrap_or_default();
    let secs = dur.as_secs();
    let tm = time::OffsetDateTime::from_unix_timestamp(secs as i64)
        .unwrap_or(time::OffsetDateTime::UNIX_EPOCH);
    format!(
        "{:04}{:02}{:02}_{:02}{:02}{:02}",
        tm.year(),
        tm.month() as u8,
        tm.day(),
        tm.hour(),
        tm.minute(),
        tm.second()
    )
}

fn dir_size(path: &PathBuf) -> u64 {
    let Ok(metadata) = std::fs::metadata(path) else {
        return 0;
    };
    if metadata.is_file() {
        return metadata.len();
    }

    let mut total = 0;
    if let Ok(entries) = std::fs::read_dir(path) {
        for entry in entries.flatten() {
            total += dir_size(&entry.path());
        }
    }
    total
}

#[cfg(test)]
mod tests {
    use super::{IncomingResumeRequest, Sandbox};
    use uuid::Uuid;

    fn temp_sandbox() -> (Sandbox, std::path::PathBuf) {
        let root = std::env::temp_dir().join(format!("syncfile-test-{}", Uuid::new_v4()));
        (Sandbox::new(root.clone()), root)
    }

    #[test]
    fn prepare_incoming_resume_returns_existing_partial_offset() {
        let (sandbox, root) = temp_sandbox();
        let file_id = "file-1";

        let request = IncomingResumeRequest {
            file_id,
            device_id: "device-1",
            device_name: "Peer",
            trust_fingerprint: "fingerprint",
            trust_public_key: "public-key",
            file_name: "photo.jpg",
            file_size: 10,
            sha256: "sha",
        };

        let first = sandbox.prepare_incoming_resume(request);
        std::fs::write(&first.partial_path, b"abc").expect("write partial file");

        let resumed = sandbox.prepare_incoming_resume(request);

        assert_eq!(resumed.bytes_received, 3);
        assert_eq!(resumed.partial_path, first.partial_path);

        let _ = std::fs::remove_dir_all(root);
    }
}
